feat/#10 게시물 상세 조회 API 구현

SteadyFlow736 opened the pull request:
## 📃 설명

- resolves #10 

## 🔨 작업 내용

1. GET /api/posts/{postId} 엔드포인트 기능 구현
2. 관련 컨트롤러, 서비스 유닛 테스트 그리고 엔드포인트 통합 테스트 추가

## 💬 기타 사항

-  Post 엔티티에서 createAt, updateAt은 외부에서 가져오는 데이터이고 우리 서비스 내에서 변화시키는 값이 아니므로 Auditing 기능을 제외하였습니다.
- 조회수 1을 추가하는 기능은 다른 feature branch에서 추가할 예정입니다.
- 권한에 따른 게시물 접근 제어는 추후 다른 feature branch에서 추가할 예정입니다.

SteadyFlow736 committed:
feat #10: 게시물 상세 조회 API 구현

- PostController에 endpoint(/api/posts/{postId} 추가
- 게시물을 가져오기 위한 PostService::getPost 메소드 추가
- 응답 Dto(GetPostResponse) 추가
- 게시물 존재하지 않을 때의 error code 추가

SteadyFlow736 committed:
feat #10: 게시물 상세 조회 API 테스트 추가

- PostController::getPost 유닛 테스트 추가
- PostService::getPost 유닛 테스트 추가
- /api/posts/{postId} endpoint 통합 테스트 추가
- PostHashtag에 Post와의 연관관계 편의 메서드 추가
- test application.yml에 sql 표시 설정 추가

mojh7 commented:
> Post 엔티티에서 createAt, updateAt은 외부에서 가져오는 데이터이고 우리 서비스 내에서 변화시키는 값이 아니므로 Auditing 기능을 제외하였습니다.

이건 따로 의심 안하고 당연하게 우리 서비스에서 Auditing 기능 이용하는거라 생각했는데
준승님이 작성한 내용이 맞는 것 같습니다.
저도 코드 작성중인 내용 Auditing 빠진 거에 맞춰서 수정해야겠네요

mojh7 commented on src/main/java/com/wanted/teamr/snsfeedintegration/service/PostService.java:
```diff
@@ -1,7 +1,23 @@
+import lombok.RequiredArgsConstructor;
 import org.springframework.stereotype.Service;
 
+@RequiredArgsConstructor
```
[[Lombok] 공식 문서를 통해 알아보는 @NoArgsConstructor, @RequiredArgsConstructor, @AllArgsConstructor + 사용시 주의사항(단점)](https://siahn95.tistory.com/170)

mojh7 replied:
`@RequiredArgsConstructor`

mojh7 commented on src/main/java/com/wanted/teamr/snsfeedintegration/dto/PostDto.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.List;
+
+@Getter
+@RequiredArgsConstructor
```
`@RequiredArgsConstructor`

mojh7 commented on src/main/java/com/wanted/teamr/snsfeedintegration/dto/GetPostResponse.java:
```diff
@@ -0,0 +1,17 @@
+import lombok.RequiredArgsConstructor;
+
+@Getter
+@RequiredArgsConstructor
```
`@RequiredArgsConstructor`

mojh7 commented on src/main/java/com/wanted/teamr/snsfeedintegration/controller/PostController.java:
```diff
@@ -1,7 +1,21 @@
+import org.springframework.web.bind.annotation.PathVariable;
 import org.springframework.web.bind.annotation.RestController;
 
+@RequiredArgsConstructor
```
`@RequiredArgsConstructor`

mojh7 reviewed (commented):
저도 예전에 봤던거라 정확하지 않아서 찾아보고 얘기하려고 수요일날 얘기 못했는데
준승님 말고도 팀원 모두와 같이 얘기하면 좋을 것 같아서 `@RequiredArgsConstructor` 있는 곳에 Comment 달았습니다.

[Lombok 사용상 주의점(Pitfall)](https://kwonnam.pe.kr/wiki/java/lombok/pitfall)
[[Lombok] 공식 문서를 통해 알아보는 @NoArgsConstructor, @RequiredArgsConstructor, @AllArgsConstructor + 사용시 주의사항(단점)](https://siahn95.tistory.com/170)

`@RequiredArgsConstructor`을 쓰면 초기화 되지 않은 모든 final fields와 선언될 때 초기화되지 않은 `@NonNull`로 표시된 field를 parameter로 가지는 생성자를 만들어준다고 하는데 이 생성자의 Parameter의 순서가 클래스 내부에서 선언된 field의 순서로 매칭 된다고 합니다.

그래서 생기는 문제가
``` java
@AllArgsConstructor
public class Order {
    private int cancelAmount;
    private int orderAmount;
}

// 취소수량 4개, 주문수량 5개
Order order = new Order(4, 5);
```
여기서 아래 처럼 필드 순서를 바꾸면
``` java
@AllArgsConstructor
public class Order {
    private int orderAmount;
    private int cancelAmount;
}
```

블로그 글에서
>이 경우, IDE가 제공해주는 리팩토링은 작동하지 않게 되고, lombok도 변화를 알아채지 못한다.
>
>심지어 orderAmount, cancelAmount는 int라는 동일한 type을 갖고 있어 더욱 버그를 잡기 어렵다.
>
>그래서 기존에 사용하던 new Order(4, 5)로 객체의 인스턴스를 만들어도 아무 에러없이 잘 동작할테지만,
>
>실제로 입력되는 값은 취소수량과 주문수량이 뒤바뀌어 들어가는 심각한 비즈니스 로직 에러를 발생시킨다.
>
>이 때문에 `@AllArgsConstructor`와 `@RequiredArgsConstructor`의 사용을 금지하는 것이 좋다는 주장도 있고,
>
>그 대신, IDE 자동 생성 기능 등으로 아래처럼 생성자를 직접 만들고,
>
>필요한 경우에는 직접 만든 생성자에 `@Builder` 어노테이션을 붙이는 것을 권장하기도 한다.
>
>이 방법은 파라미터 순서가 아닌 이름으로 값을 설정하기 때문에 리팩토링에 유연하게 대응이 가능하다.

라고 해서 저도 처음에 `@RequiredArgsConstructor`, `@AllArgsConstructor` 썼었는데
지금은 둘 다 안쓰고 생성자 주입만 사용하고 있습니다. (직접 적으면 번거로워서 인텔리제이 윈도우 기준 Alt+Insert Generate 기능으로 생성합니다)
entity에 경우 생성자에다가 입력해도 될 필드만 남기고 `@Builder`만 붙여서 쓰고 있습니다.

혹시 위에 링크에 나온 내용이랑 다르게 따로 알고 계신 내용이나 `@RequiredArgsConstructor`를 써도 문제없는 이유가 따로 있을까요??

mojh7 commented on src/main/java/com/wanted/teamr/snsfeedintegration/dto/PostDto.java:
```diff
@@ -0,0 +1,36 @@
+    private final LocalDateTime updatedAt;
+    private final LocalDateTime createdAt;
+
+    public static PostDto of(Post post) {
```
이건 개인적인 취향인지 모르겠는데 저도 Response쪽 dto 객체 생성할 때 of(...), from(...) 이런식으로 정적 팩토리 메서드 패턴으로 만드는 것 선호합니다.

SteadyFlow736 commented:
> 저도 예전에 봤던거라 정확하지 않아서 찾아보고 얘기하려고 수요일날 얘기 못했는데 준승님 말고도 팀원 모두와 같이 얘기하면 좋을 것 같아서 `@RequiredArgsConstructor` 있는 곳에 Comment 달았습니다.
> 
> [Lombok 사용상 주의점(Pitfall)](https://kwonnam.pe.kr/wiki/java/lombok/pitfall) [[Lombok] 공식 문서를 통해 알아보는 @NoArgsConstructor, @RequiredArgsConstructor, @AllArgsConstructor + 사용시 주의사항(단점)](https://siahn95.tistory.com/170)
> 
> `@RequiredArgsConstructor`을 쓰면 초기화 되지 않은 모든 final fields와 선언될 때 초기화되지 않은 `@NonNull`로 표시된 field를 parameter로 가지는 생성자를 만들어준다고 하는데 이 생성자의 Parameter의 순서가 클래스 내부에서 선언된 field의 순서로 매칭 된다고 합니다.
> 
> 그래서 생기는 문제가
> 
> ```java
> @AllArgsConstructor
> public class Order {
>     private int cancelAmount;
>     private int orderAmount;
> }
> 
> // 취소수량 4개, 주문수량 5개
> Order order = new Order(4, 5);
> ```
> 
> 여기서 아래 처럼 필드 순서를 바꾸면
> 
> ```java
> @AllArgsConstructor
> public class Order {
>     private int orderAmount;
>     private int cancelAmount;
> }
> ```
> 
> 블로그 글에서
> 
> > 이 경우, IDE가 제공해주는 리팩토링은 작동하지 않게 되고, lombok도 변화를 알아채지 못한다.
> > 심지어 orderAmount, cancelAmount는 int라는 동일한 type을 갖고 있어 더욱 버그를 잡기 어렵다.
> > 그래서 기존에 사용하던 new Order(4, 5)로 객체의 인스턴스를 만들어도 아무 에러없이 잘 동작할테지만,
> > 실제로 입력되는 값은 취소수량과 주문수량이 뒤바뀌어 들어가는 심각한 비즈니스 로직 에러를 발생시킨다.
> > 이 때문에 `@AllArgsConstructor`와 `@RequiredArgsConstructor`의 사용을 금지하는 것이 좋다는 주장도 있고,
> > 그 대신, IDE 자동 생성 기능 등으로 아래처럼 생성자를 직접 만들고,
> > 필요한 경우에는 직접 만든 생성자에 `@Builder` 어노테이션을 붙이는 것을 권장하기도 한다.
> > 이 방법은 파라미터 순서가 아닌 이름으로 값을 설정하기 때문에 리팩토링에 유연하게 대응이 가능하다.
> 
> 라고 해서 저도 처음에 `@RequiredArgsConstructor`, `@AllArgsConstructor` 썼었는데 지금은 둘 다 안쓰고 생성자 주입만 사용하고 있습니다. (직접 적으면 번거로워서 인텔리제이 윈도우 기준 Alt+Insert Generate 기능으로 생성합니다) entity에 경우 생성자에다가 입력해도 될 필드만 남기고 `@Builder`만 붙여서 쓰고 있습니다.
> 
> 혹시 위에 링크에 나온 내용이랑 다르게 따로 알고 계신 내용이나 `@RequiredArgsConstructor`를 써도 문제없는 이유가 따로 있을까요??

이 문제는 인식하지 못하고 있었네요. 좋은 정보 감사 드립니다.
정리하자면, @AllArgsConstructor @RequiredArgsConstructor는 클래스 내부에 명명된 field 의 순서를 constructor가 호출될 때마다 확인해서 constructor의 parameter 위치를 바꿔주는데 반해,
우리가 작성한 코드는 argument가 그대로이므로 argument가 parameter에 잘못 연결이 될 수 있다는 것이군요!

회의록에서 정한대로 가면 될 것 같습니다.
@RestController, @Service 의 constructor를 프레임워크에서 호출하므로 argument가 잘못 전달될 리는 없을 것이므로 그 둘은 @RequiredArgsConstructor, @AllArgsConstructor를 사용해도 될 것 같습니다.
반면 Dto, Entity의 constructor는 우리가 호출하고 argument를 전달하므로 실수의 여지가 있으므로 Dto에는 저 두개의 annotation을 사용 안하는 게 맞는 결정인 것 같습니다.

mojh7 reviewed (approved):
제가 작성했던 내용 대부분 회의에서 의견 나눴던 것들이라 리뷰 열린거 닫았고
PR approved 처리했습니다. 고생하셨습니다 👍👍👍

SteadyFlow736 committed:
feat #10: 컨벤션에 따른 게시물 상세 조회 API 수정

- Controller의 Response DTO를 service에서 직접 리턴하도록 변경
- Entity는 기본적으로 private 생성자에 @Builder를 사용하도록 변경
- 유닛 테스트의 suffix를 MockTest, 통합 테스트의 suffix를 Test 로 변경하고 패키지는 프로덕션과 동일한 위치로 변경

SteadyFlow736 commented:
회의록에서 정한 컨벤션을 지키도록 수정한 커밋 올렸습니다!

piopoi commented on src/test/java/com/wanted/teamr/snsfeedintegration/controller/PostControllerMockTest.java:
```diff
@@ -0,0 +1,88 @@
+    @DisplayName("게시물 상세 정보 응답을 보낸다.")
+    @WithMockUser
+    @Test
+    void getPost() throws Exception {
```
궁금한 부분이 있어 질문남깁니다 ✋ 
`PostControllerMockTest.getPost()` 와 `PostControllerTest.getPost()`는 테스트 라이브러리만 다를 뿐 같은 역할로 보이는데, 2가지 모두 작성하신 이유가 있으실까요?

piopoi replied:
코드리뷰 시간에 충분히 답변을 해주셔서 approve 하겠습니다! 😄

SteadyFlow736 replied:
이런 것들이 막상 설명하려고 하면 잘 안되는 것들인 것 같습니다.
질문 주셔서 이번 기회에 관련하여 생각을 정리할 수 있었습니다. 감사합니다! 😀

piopoi reviewed (approved):
코딩 컨벤션이랑 요구사항 잘 지키셨네요.
수고하셨습니다. 😄
